Lib/asyncio/locks.py

```diff
@@ -418,3 +418,174 @@ def release(self):
+# similar to 'draining' except that tasks leave with a BrokenBarrierError,
+# and a 'broken' state in which all tasks get the exception.
+
+class Barrier(mixins._LoopBoundMixin):
+    """Asynchronous equivalent to threading.Barrier
+
+    Implements a Barrier.
+    Useful for synchronizing a fixed number of tasks at known synchronization
+    points.  Tasks block on 'wait()' and are simultaneously awoken once they
+    have all made that call.
+    """
+
+    def __init__(self, parties, action=None, *, loop=mixins._marker):
+        """Create a barrier, initialised to 'parties' tasks.
+        'action' is a callable which, when supplied, will be called by one of
+        the tasks after they have all entered the barrier and just prior to
+        releasing them all.
+        """
+        super().__init__(loop=loop)
+        if parties < 1:
+            raise ValueError('parties must be > 0')
+
+        self._waiting = Event()     # used notify all waiting tasks
+        self._blocking = Event()    # used block new tasks while waiting tasks are draining or broken
+        self._action = action
+        self._parties = parties
+        self._state = 0             # 0 filling, 1, draining, -1 resetting, -2 broken
+        self._count = 0             # count waiting tasks
+
+    def __repr__(self):
+        res = super().__repr__()
+        _wait = 'set' if self._waiting.is_set() else 'unset'
+        _block = 'set' if self._blocking.is_set() else 'unset'
+        extra = f'{_wait}, count:{self._count}/{self._parties}, {_block}, state:{self._state}'
+        return f'<{res[1:-1]} [{extra}]>'
+
+    async def wait(self):
+        """Wait for the barrier.
+        When the specified number of tasks have started waiting, they are all
+        simultaneously awoken. If an 'action' was provided for the barrier, one
+        of the tasks will have executed that callback prior to returning.
+        Returns an individual index number from 0 to 'parties-1'.
+        """
+        await self._block() # Block while the barrier drains or resets.
+        index = self._count
+        self._count += 1
+        try:
+            if index + 1 == self._parties:
+                # We release the barrier
+                self._release()
+            else:
+                # We wait until someone releases us
+                await self._wait()
+            return index
+        finally:
+            self._count -= 1
+            # Wake up any tasks waiting for barrier to drain.
+            self._exit()
+
+    # Block until the barrier is ready for us, or raise an exception
+    # if it is broken.
+    async def _block (self):
+        while self._state in (-1, 1):
+            # It is draining or resetting, wait until done
+            await self._blocking.wait()
+
+        #see if the barrier is in a broken state
+        if self._state < 0:
+            raise BrokenBarrierError
+        assert self._state == 0, repr(self)
```

Comment:
Ok, this ``assert`` and this one of ``_wait`` method will be removed. 
In ``_block`` method, it is not necessary to add a new test because ``self._state`` is always equal to 0
In ``_wait`` method, no more tests are needed